docs/usage.rst

```diff
@@ -0,0 +1,98 @@
+*****
+Usage
+*****
+
+Run Hyperparameter Search
+=========================
+
+cluster_utils provides two main commands to run hyperparameter search in different
+modes:
+
+- ``grid_search``:  Simple grid search over specified hyperparameter ranges.
+- ``hp_optimization``:  Uses sampling-based optimization to search for best combination
+  of hyperparameters within the specified ranges.
+
+They are both run as modules via ``python -m`` and expect a configuration file as
+argument:
+
+.. code-block:: bash
+
+   python3 -m cluster_utils.grid_search config_for_grid_search.json
+
+   python3 -m cluster_utils.hp_optimization config_for_hp_optim.json
+
+See :doc:`configuration` for information on the expected structure of the config file
+(note that there are differences between the two methods!).
+
+Optionally, a list of key-value arguments can be provided in addition, to overwrite
+single settings from the config file.  Use dot-notation to specify nested keys.  Example:
+
+.. code-block:: bash
+
+   python3 -m cluster_utils.grid_search config.json \
+     'results_dir="/tmp"' 'optimization_setting.run_local=True'
+
+Both commands can also be run with ``--help`` to get a complete list of arguments.
+
+You can abort the hyperparameter search with Ctrl + C at any time.
+
+
+Interactive Mode
+================
+
+While cluster_utils is running, it is possible to enter a command prompt which allows to
+get some information about finished and running jobs, as well as to stop running jobs.
+
+To enter the command prompt, press ESC.  You should now see the following prompt:
+
+::
+
+   Enter command, e.g.  list_jobs, list_running_jobs, list_successful_jobs,
+   list_idle_jobs, show_job, stop_remaining_jobs
+   >>>
+
+You now may enter one of the listed commands or simply press Enter to leave the prompt
+without executing a command.
+
+.. important::
+
+   While the prompt is open, the main loop is blocked, i.e. no new jobs will be
+   submitted during that time.
+
+
+Commands
+--------
+
+.. I'm a bit misusing the confval directive here, but I think as long as there is no
+   name collision with an actual config value, this should be fine and much easier than
+   adding a dedicated directive.
+
+.. confval:: list_jobs
+
+   List IDs of all jobs that have been submitted so far (including finished ones).
+
+.. confval:: list_running_jobs
+
+   List IDs of all jobs that are currently running.
+
+.. confval:: list_successful_jobs
+
+   List IDs of all jobs that finished successfully.
+
+.. confval:: list_idle_jobs
+
+   List IDs of all jobs that have been submitted but not yet started.
+
+.. confval:: show_job
+
+   Will ask for a job ID and show information about this job.
+
+.. confval:: stop_remaining_jobs
+
+   Abort all currently running jobs as well as jobs that already have been submitted but
+   didn't start yet.
+
+   This will not stop submission of new jobs.  If you want to stop cluster_utils
```

Comment:
It does not stop cluster_utils fully?  Is this the case for hp_optimization? If so, this is a bug, because on grid_search, it does (and is the intended outcome of the command).

Reply:
Yes, this is the case for hp_optimization.  I think the difference to grid_search is that there all jobs will be submitted at once while hp_optimization waits for first jobs to finish before submitting the next ones.
From what I understood, the command only cancels running and submitted jobs but doesn't exit the main loop, so once they are cancelled, it proceeds submitting the next batch.  I have to admit, I was a bit confused about this behaviour back when I noticed it :D.

Another funny thing I just noticed: After calling it, I get a rather high number of failed jobs displayed in the "started execution" progress bar which then goes *down* over the next few seconds.  So also something is wrongly counted there.

Anyway, if the purpose is to stop everything, do we actually need that command?  Does it have any advantage over stopping with Ctrl+C?

Reply:
I created issue #116 for this.